cvatjs/src/annotations.js

```diff
@@ -287,73 +336,402 @@
+
+                for (let i = 0; i < points.length; i += 2) {
+                    normalized.push(
+                        (points[i] - box.xmin) / width,
+                        (points[i + 1] - box.ymin) / height,
+                    );
+                }
+
+                return normalized;
             }
 
-            if (rightPosition) {
-                return {
-                    points: [...rightPosition.points],
-                    occluded: rightPosition.occluded,
-                    outside: true,
-                    zOrder: 0,
-                };
+            function denormalize(points, box) {
+                const denormalized = [];
+                const width = box.xmax - box.xmin;
+                const height = box.ymax - box.ymin;
+
+                for (let i = 0; i < points.length; i += 2) {
+                    denormalized.push(
+                        points[i] * width + box.xmin,
+                        points[i + 1] * height + box.ymin,
+                    );
+                }
+
+                return denormalized;
             }
 
-            if (leftPosition) {
-                return {
-                    points: [...leftPosition.points],
-                    occluded: leftPosition.occluded,
-                    outside: leftPosition.outside,
-                    zOrder: 0,
+            function toPoints(array) {
+                const points = [];
+                for (let i = 0; i < array.length; i += 2) {
+                    points.push({
+                        x: array[i],
+                        y: array[i + 1],
+                    });
+                }
+
+                return points;
+            }
+
+            function toArray(points) {
+                const array = [];
+                for (const point of points) {
+                    array.push(point.x, point.y);
+                }
+
+                return array;
+            }
+
+            function computeDistances(source, target) {
+                const distances = {};
+                for (let i = 0; i < source.length; i++) {
+                    distances[i] = distances[i] || {};
+                    for (let j = 0; j < target.length; j++) {
+                        const dx = source[i].x - target[j].x;
+                        const dy = source[i].y - target[j].y;
+
+                        distances[i][j] = Math.sqrt(Math.pow(dx, 2) + Math.pow(dy, 2));
+                    }
+                }
+
+                return distances;
+            }
+
+            function truncateByThreshold(mapping, threshold) {
+                for (const key of Object.keys(mapping)) {
+                    if (mapping[key].distance > threshold) {
+                        delete mapping[key];
+                    }
+                }
+            }
+
+            // https://en.wikipedia.org/wiki/Stable_marriage_problem
+            function stableMarriageProblem(men, women, distances) {
```

Comment:
One of important part of the algorithm is to correctly match "corner" points. Thus it is possible for each of such point calculate a descriptor (d) and use (x, y, d) to calculate the distance. One more idea is to be sure that order or matched points is preserved. For example, if p1 matches q1 and p2 matches q2 and between p1 and p2 we don't have any points thus we should not have points between p2 and q2 as well.